Frameworks/UIKit/UIGraphicsFunctions.mm

```diff
@@ -101,63 +126,91 @@ void UIGraphicsBeginImageContext(CGSize size) {
  @Status Interoperable
 */
 UIImage* UIGraphicsGetImageFromCurrentImageContext() {
-    return nil;
-    // TODO(DH) GH#1124 Revisit when we have support for CGImage+IWICBitmap
-#if 0
-    id ret = [UIImage imageWithCGImage:CGBitmapContextGetImage(UIGraphicsGetCurrentContext())
-                                 scale:UIGraphicsGetCurrentContext()->scale
-                           orientation:UIImageOrientationUp];
+    std::lock_guard<std::mutex> lock(s_contextStackMutex);
+    if (s_contextStack.size() == 0) {
+        TraceError(TAG, L"UIGraphicsGetImageFromCurrentImageContext(): the context stack was empty.");
+        return nil;
+    }
+
+    _UIGraphicsContextRecord& topRecord = s_contextStack.top();
+    if (topRecord.type != _UIGraphicsContextTypeImage) {
+        TraceError(TAG, L"UIGraphicsGetImageFromCurrentImageContext(): the top context is not an image context.");
+        return nil;
+    }
 
+    CGContextRef context = topRecord.context.get();
+    id ret = [UIImage imageWithCGImage:CGBitmapContextGetImage(context) scale:topRecord.scale orientation:UIImageOrientationUp];
     return ret;
-#endif
 }
 
 /**
  @Status Interoperable
 */
 void UIGraphicsEndImageContext() {
-    CGContextRef ctx = UIGraphicsGetCurrentContext();
+    std::lock_guard<std::mutex> lock(s_contextStackMutex);
+    if (s_contextStack.size() == 0) {
+        TraceError(TAG, L"UIGraphicsEndImageContext(): the context stack was empty.");
+        return;
+    }
+
+    _UIGraphicsContextRecord& topRecord = s_contextStack.top();
+    if (topRecord.type != _UIGraphicsContextTypeImage) {
+        return;
+    }
 
-    UIGraphicsPopContext();
-    CGContextRelease(ctx);
+    s_contextStack.pop(); // Releases record's CGContext.
 }
 
 /**
  @Status Interoperable
 */
 void UIRectFill(CGRect rect) {
-    CGContextRef ctx = UIGraphicsGetCurrentContext();
-    CGContextFillRect(ctx, rect);
+    UIRectFillUsingBlendMode(rect, kCGBlendModeCopy);
 }
 
 /**
- @Status Stub
+ @Status Caveat
+ @Notes Blend modes are not implemented.
 */
-void UIRectFrame(CGRect rect) {
-    UNIMPLEMENTED();
-    CGContextRef ctx = UIGraphicsGetCurrentContext();
-    // CGContextFillRect(ctx, rect);
-    TraceVerbose(TAG, L"UIRectFrame not supported");
+void UIRectFillUsingBlendMode(CGRect rect, CGBlendMode blendMode) {
+    CGContextRef context = UIGraphicsGetCurrentContext();
+
+    CGContextSaveGState(context);
+    CGContextSetBlendMode(context, blendMode);
+    CGContextFillRect(context, rect);
+    CGContextRestoreGState(context);
 }
 
 /**
  @Status Interoperable
 */
-void UIRectClip(CGRect clip) {
-    CGContextRef ctx = UIGraphicsGetCurrentContext();
+void UIRectFrame(CGRect rect) {
+    UIRectFrameUsingBlendMode(rect, kCGBlendModeCopy);
+}
 
-    CGContextClipToRect(ctx, clip);
+/**
+ @Status Caveat
+ @Notes Blend modes are not implemented.
```

Comment:
Since these caveats are already present in CG, we don't need to specify it again here.  I would mark these as interoperable with notes that we support all blend modes that CG supports.